File: software/protocol/commands.py
"""Safe parsing and formatting for the NEXO serial protocol."""
from enum import Enum

class Command(str, Enum):
    STOP = "STOP"
    FORWARD = "FORWARD"
    BACK = "BACK"
    LEFT = "LEFT"
    RIGHT = "RIGHT"
    STATUS = "STATUS"
    PING = "PING"

ALLOWED_COMMANDS = frozenset(Command)

def parse_command(raw: str) -> Command | None:
    if not isinstance(raw, str):
        return None
    try:
        return Command(raw.strip().upper())
    except ValueError:
        return None

def format_command(command: Command) -> str:
    if command not in ALLOWED_COMMANDS:
        raise ValueError("Command is not allowlisted")
    return command.value + "\n"

File: software/arduino/commands.cpp
#include "commands.h"

RobotCommand parseCommand(String input) {
  input.trim();
  input.toUpperCase();

  if (input == "STOP") return RobotCommand::Stop;
  if (input == "FORWARD" || input == "F") return RobotCommand::Forward;
  if (input == "BACK" || input == "BACKWARD" || input == "B") return RobotCommand::Back;
  if (input == "LEFT" || input == "L") return RobotCommand::Left;
  if (input == "RIGHT" || input == "R") return RobotCommand::Right;
  if (input == "STATUS" || input == "S") return RobotCommand::Status;

  return RobotCommand::Unknown;
}

const __FlashStringHelper* commandName(RobotCommand command) {
  switch (command) {
    case RobotCommand::Stop: return F("STOP");
    case RobotCommand::Forward: return F("FORWARD");
    case RobotCommand::Back: return F("BACK");
    case RobotCommand::Left: return F("LEFT");
    case RobotCommand::Right: return F("RIGHT");
    case RobotCommand::Status: return F("STATUS");
    default: return F("UNKNOWN");
  }
}
